trpcultivate_phenotypes/tests/src/Kernel/TripalImporter/TraitImporterFormValidateTest.php

```diff
@@ -84,122 +84,228 @@ protected function setUp(): void {
-   *      each unique validator instance that was instantiated by the 
+   *    - Each key is the unique name of a feedback line provided to the validator UI
+   *      by the processValidationMessages(). Currently there is a feedback line for
+   *      each unique validator instance that was instantiated by the
    *      configureValidators() method in the Traits Importer class
    *      - 'status': [REQUIRED] One of 'pass', 'todo', or 'fail'
    *      - 'title': [REQUIRED if 'status' = 'fail'] A string that matches the
    *        title set in processValidationMessages() method in the Trait Importer
    *        class for this validator instance.
    *      - 'details': [REQUIRED if 'status' = 'fail'] A string that is passed
    *        to the user through the UI if this validation instance failed.
+   * -- an integer of the number of form validation messages we expect to see
+   *    when the form is submitted. NOTE: These validation messages are produced
+   *    by the form via Drupal and are not related to this module's use of
+   *    validator plugins.
    */
   public function provideFilesForValidation() {
-    $senarios = [];
 
-    // #0: Contains correct header but no data
+    $valid_genus = 'Tripalus';
+    $invalid_genus = 'INVALID';
+    // Set our number of expected validation messages to 0, since only the
+    // 'genus_exists' validator should cause this number to change.
+    $num_form_validation_messages = 0;
+
+    $scenarios = [];
+
+    // #0: File is valid but genus is not
+    $scenarios[] = [
+      $invalid_genus,
+      'simple_example.txt',
+      [
+        'genus_exists' => [
+          'title' => 'The genus is valid',
+          'status' => 'fail',
+          'details' => 'Genus does not exist'
+        ],
+        'valid_data_file' => ['status' => 'todo'],
+        'valid_delimited_file' => ['status' => 'todo'],
+        'valid_header' => ['status' => 'todo'],
+        'empty_cell' => ['status' => 'todo'],
+        'valid_data_type' => ['status' => 'todo'],
+        'duplicate_traits' => ['status' => 'todo']
+      ],
+      1 // Since selecting an invalid genus should be impossible, 1 form validation error is expected
+    ];
+
+    // #1: File is empty
+    $scenarios[] = [
+      $valid_genus,
+      'empty_file.txt',
+      [
+        'genus_exists' => ['status' => 'pass'],
+        'valid_data_file' => [
+          'title' => 'File is valid and not empty',
+          'status' => 'fail',
+          'details' => 'The file has no data and is an empty file'
+        ],
+        'valid_delimited_file' => ['status' => 'todo'],
+        'valid_header' => ['status' => 'todo'],
+        'empty_cell' => ['status' => 'todo'],
+        'valid_data_type' => ['status' => 'todo'],
+        'duplicate_traits' => ['status' => 'todo']
+      ],
+      $num_form_validation_messages
+    ];
+
+    // #2: 2nd row of file is improperly delimited
```

Comment:
It would be really good to have a test scenario where the header row is improperly delimited. Just to confirm raw row is also functioning on the header and that all data row validation is skipped if the header is bad. I would use a file with two correct data rows and a badly delimited but otherwise perfect header.

Reply:
Great idea - scenario added in commit [56cae6a](https://github.com/TripalCultivate/TripalCultivate-Phenotypes/pull/108/commits/56cae6ae718a2e54e38e1110c92912911d3707d6) 😊